File: golang/7.sort/insertion_sort.go
package main

import "reflect"

// o(N^2)
func InsertionSort(arr *[]int) {
    sf := reflect.Swapper(*arr) ; 

    if len(*arr) < 2 { 
        return ; 
    }

    for i:=0 ; i < len(*arr) ; i = i + 1 { 
        KI := i + 1 ; 
        if KI >= len(*arr) { 
            break ;
        }
        K := (*arr)[KI] ;

        for j := KI - 1 ; j >= 0; j = j - 1 { 
            L := (*arr)[j] ; 

            if L > K { 
                sf(KI , j) ; 
                KI = KI - 1 ; 
                K = (*arr)[KI] ; 
            }
        }
    }
}

File: ts/6.sort/insertion-sort/insertion-sort.ts
// o(N^2) 
export function insertionSort(arr: number[]) {
    if (arr.length < 2) return;

    for (let j = 0; j < arr.length; j++) {
        let keyIndex = j + 1
        if (keyIndex >= arr.length) break;

        let key = arr[keyIndex];

        for (let i = keyIndex - 1; i >= 0; i--) {
            let L = arr[i];

            if (L > key) {
                const temp = arr[i];
                arr[keyIndex] = temp;
                arr[i] = key;

                // change the key 
                keyIndex = keyIndex - 1;
                key = arr[keyIndex];
            }
        }
    }
}
